src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs

```diff
@@ -351,6 +351,279 @@ static S1 MayWrap(Span<int> arg)
+                // (32,20): error CS8347: Cannot use a result of 'Program.ReturnsRef1(__arglist)' in this context because it may expose variables referenced by parameter '__arglist' outside of their declaration scope
+                //         return ref ReturnsRef1(__arglist(ref local));
+                Diagnostic(ErrorCode.ERR_EscapeCall, "ReturnsRef1(__arglist(ref local))").WithArguments("Program.ReturnsRef1(__arglist)", "__arglist").WithLocation(32, 20)
+            );
+        }
+
+        [Fact()]
+        public void ThrowExpression()
+        {
+            var text = @"
+using System;
+class Program
+{
+    static void Main()
+    {
+
+    }    
+
+    static ref int M1() => throw null;
+
+    static ref readonly int M2() => throw null;
+
+    static ref Span<int> M3() => throw null;
+
+    static ref readonly Span<int> M4() => throw null;
+
+    static Span<int> M5() => throw null;
+
+    static Span<int> M6() => M5().Length !=0 ? M5() : throw null;
+
+}
+";
+            CreateCompilationWithMscorlibAndSpan(text).VerifyDiagnostics();
+        }
+
+
+        [Fact()]
+        public void UserDefinedLogical()
+        {
+            var text = @"
+using System;
+class Program
+{
+    static void Main()
+    {
+    }
+
+    S1 Test()
+    {
+        S1 global = default;
+        S1 local = stackalloc int[100];
+
+        // ok
+        local = global && local;
+        local = local && local;
+
+        // ok
+        global = global && global;
+
+        // error
+        global = local && global;
+
+        // error
+        return global || local;
+    }
+}
+
+ref struct S1
+{
+    public static implicit operator S1(Span<int> o) => default;
+
+    public static bool operator true(S1 o) => true;
+    public static bool operator false(S1 o) => false;
+
+    public static S1 operator &(S1 x, S1 y) => x;
+    public static S1 operator |(S1 x, S1 y) => x;
+}
+
+";
+            CreateCompilationWithMscorlibAndSpan(text).VerifyDiagnostics(
+                // (22,18): error CS8352: Cannot use local 'local' in this context because it may expose referenced variables outside of their declaration scope
+                //         global = local && global;
+                Diagnostic(ErrorCode.ERR_EscapeLocal, "local").WithArguments("local").WithLocation(22, 18),
+                // (25,26): error CS8352: Cannot use local 'local' in this context because it may expose referenced variables outside of their declaration scope
+                //         return global || local;
+                Diagnostic(ErrorCode.ERR_EscapeLocal, "local").WithArguments("local").WithLocation(25, 26)
+             );
+        }
+
+        [Fact()]
+        public void DiscardExpressionRef()
+        {
+            var text = @"
+
+class Program
+{
+    static void Main()
+    {
+
+    }
+
+    static ref int ReturnsRef()
+    {
+        return ref ReturnsRef1(out var _ );
```

Comment:
If it's not tested already, could you also test the same scenario with a named local? Should behave the same.
Also, nit: space after discard